sql: TestRelocateNonVoters failed

sql.TestRelocateNonVoters [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21332654?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21332654?buildTab=artifacts#/) on master @ [f244f91f126e0fb64a2957e5fa628486c6eec553](https://github.com/cockroachdb/cockroach/commits/f244f91f126e0fb64a2957e5fa628486c6eec553):


```
=== RUN   TestRelocateNonVoters
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/37acaf1de75f910fa45ac01dbd64243a/logTestRelocateNonVoters2526354872
    test_log_scope.go:82: use -show-logs to present logs inline
    multitenant_admin_function_test.go:1011: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37acaf1de75f910fa45ac01dbd64243a/logTestRelocateNonVoters2526354872
--- FAIL: TestRelocateNonVoters (85.78s)
=== RUN   TestRelocateNonVoters/ALTER_RANGE_x_RELOCATE_NONVOTERS
    multitenant_admin_function_test.go:138: condition failed to evaluate within 45s: from multitenant_admin_function_test.go:125: numReplicas expected=4 actual=5 tenant=secondary query=``ALTER RANGE (SELECT min(range_id) FROM [SHOW RANGES FROM TABLE t]) RELOCATE NONVOTERS FROM 1 TO 2;`` leaseholder=4 replicas=[1 3 4 5] voting_replicas=[3 4 5] non_voting_replicas=[1] fromReplica=1 toReplica=2
        
        goroutine dump: /artifacts/tmp/_tmp/37acaf1de75f910fa45ac01dbd64243a/goroutine_dump_2792042577.txt
--- FAIL: TestRelocateNonVoters/ALTER_RANGE_x_RELOCATE_NONVOTERS (64.60s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRelocateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63419

--- a/pkg/sql/multitenant_admin_function_test.go
+++ b/pkg/sql/multitenant_admin_function_test.go
@@ -167,6 +167,58 @@ FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS]`)
 	}
 }
 
+// disableQueuesAndStabilize disables the split, lease, and replicate queues and
+// waits for replica state to stabilize.
+func disableQueuesAndStabilize(
+	t *testing.T,
+	ctx context.Context,
+	testCluster serverutils.TestClusterInterface,
+	db *gosql.DB,
+	expectedNumReplicas int,
+) {
+	// Prevent new splits and lease transfers.
+	testCluster.ToggleSplitQueues(false)
+	testCluster.ToggleLeaseQueues(false)
+
+	// Wait for learner replicas to be promoted. The replicate queue must stay
+	// active throughout because learner promotion requires a Raft snapshot
+	// transfer that cannot complete if the queue is toggled on and off.
+	testutils.SucceedsSoon(t, func() error {
+		var count int
+		if err := db.QueryRowContext(ctx,
+			`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS] WHERE learner_replicas != '{}'`,
+		).Scan(&count); err != nil {
+			return err
+		}
+		if count > 0 {
+			return errors.Newf("waiting for %d ranges with LEARNER replicas to clear", count)
+		}
+		return nil
+	})
+
+	// Disable the replicate queue and check that replica counts are correct.
+	// If a rebalance was in progress (add phase done, remove phase pending),
+	// re-enable the queue so it can finish, then retry.
+	testutils.SucceedsSoon(t, func() error {
+		testCluster.ToggleReplicateQueues(false)
+		var count int
+		if err := db.QueryRowContext(ctx,
+			fmt.Sprintf(
+				`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS]
+				 WHERE array_length(replicas, 1) != %d`, expectedNumReplicas),
+		).Scan(&count); err != nil {
+			return err
+		}
+		if count > 0 {
+			testCluster.ToggleReplicateQueues(true)
+			return errors.Newf(
+				"waiting for %d ranges with unexpected replica count to stabilize", count,
+			)
+		}
+		return nil
+	})
+}
+
 // getToReplica determines the "to" node for a relocate command by finding the NodeID
 // in clusterNodeIDs that is not in replicaNodeIDs.
 func getToReplica(clusterNodeIDs []roachpb.NodeID, replicaNodeIDs []roachpb.NodeID) roachpb.NodeID {
@@ -768,21 +820,7 @@ func TestRelocateVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
-					testutils.SucceedsSoon(t, func() error {
-						var count int
-						if err := db.QueryRowContext(ctx,
-							`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS] WHERE learner_replicas != '{}'`,
-						).Scan(&count); err != nil {
-							return err
-						}
-						if count > 0 {
-							return errors.Newf("waiting for %d ranges with LEARNER replicas to clear", count)
-						}
-						return nil
-					})
-					testCluster.ToggleLeaseQueues(false)
-					testCluster.ToggleReplicateQueues(false)
-					testCluster.ToggleSplitQueues(false)
+					disableQueuesAndStabilize(t, ctx, testCluster, db, expectedNumReplicas)
 					tExp.validate(
 						t,
 						func() (_ *gosql.Rows, msg string, _ error) {
@@ -861,21 +899,7 @@ func TestExperimentalRelocateVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
-					testutils.SucceedsSoon(t, func() error {
-						var count int
-						if err := db.QueryRowContext(ctx,
-							`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS] WHERE learner_replicas != '{}'`,
-						).Scan(&count); err != nil {
-							return err
-						}
-						if count > 0 {
-							return errors.Newf("waiting for %d ranges with LEARNER replicas to clear", count)
-						}
-						return nil
-					})
-					testCluster.ToggleLeaseQueues(false)
-					testCluster.ToggleReplicateQueues(false)
-					testCluster.ToggleSplitQueues(false)
+					disableQueuesAndStabilize(t, ctx, testCluster, db, expectedNumReplicas)
 					tExp.validate(
 						t,
 						func() (*gosql.Rows, string, error) {
@@ -967,21 +991,7 @@ func TestRelocateNonVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
-					testutils.SucceedsSoon(t, func() error {
-						var count int
-						if err := db.QueryRowContext(ctx,
-							`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS] WHERE learner_replicas != '{}'`,
-						).Scan(&count); err != nil {
-							return err
-						}
-						if count > 0 {
-							return errors.Newf("waiting for %d ranges with LEARNER replicas to clear", count)
-						}
-						return nil
-					})
-					testCluster.ToggleLeaseQueues(false)
-					testCluster.ToggleReplicateQueues(false)
-					testCluster.ToggleSplitQueues(false)
+					disableQueuesAndStabilize(t, ctx, testCluster, db, expectedNumReplicas)
 					tExp.validate(
 						t,
 						func() (*gosql.Rows, string, error) {
@@ -1054,21 +1064,7 @@ func TestExperimentalRelocateNonVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
-					testutils.SucceedsSoon(t, func() error {
-						var count int
-						if err := db.QueryRowContext(ctx,
-							`SELECT count(*) FROM [SHOW RANGES FROM INDEX t@primary WITH DETAILS] WHERE learner_replicas != '{}'`,
-						).Scan(&count); err != nil {
-							return err
-						}
-						if count > 0 {
-							return errors.Newf("waiting for %d ranges with LEARNER replicas to clear", count)
-						}
-						return nil
-					})
-					testCluster.ToggleLeaseQueues(false)
-					testCluster.ToggleReplicateQueues(false)
-					testCluster.ToggleSplitQueues(false)
+					disableQueuesAndStabilize(t, ctx, testCluster, db, expectedNumReplicas)
 					tExp.validate(
 						t,
 						func() (*gosql.Rows, string, error) {

